Add determinant methods in solve module

jturner314 opened the pull request:
This implementation is based on the [NumPy implementation](https://github.com/numpy/numpy/blob/v1.13.1/numpy/linalg/umath_linalg.c.src#L976-L1131). It uses the decomposition `A = P * L * U => det(A) = det(P) * det(L) * det(U)`, where `det(P)` is ±1 based on the number of pivots, `det(L)` is 1, and `det(U)` is the product of the diagonal elements of `U`.

termoshtt commented on src/lapack_traits/solve.rs:
```diff
@@ -10,6 +10,14 @@ use super::{Pivot, Transpose, into_result};
+    /// Computes the LU factorization of a general `m x n` matrix `a` using
+    /// partial pivoting with row interchanges.
+    ///
+    /// If the result matches `Err(LinalgError::Lapack(LapackError {
```
It is fortran order. Hence `U[(return_code-1, return_code-1)]` is zero.

termoshtt commented on src/solve.rs:
```diff
@@ -328,3 +328,100 @@ where
+    pivot_sign * upper_sign * ln_det.exp()
+}
+
+fn check_square<S: Data>(a: &ArrayBase<S, Ix2>) -> Result<()> {
```
move to `layout` module, and make `pub` since it will be useful.

jturner314 commented:
> It is fortran order. Hence `U[(return_code-1, return_code-1)]` is zero.

Good catch. Fixed.

> move to `layout` module, and make `pub` since it will be useful.

I moved it to the `AllocatedArray` trait since that seemed like an appropriate place. If you'd prefer a standalone function, I can do that instead.

termoshtt commented:
> I moved it to the AllocatedArray trait since that seemed like an appropriate place.

Good choice :)

Thanks a lot!
